builder/etcd/etcd.go

```diff
@@ -0,0 +1,529 @@
+/*Package etcd is a library for performing common Etcd tasks.
+ */
+package etcd
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/Masterminds/cookoo"
+	"github.com/Masterminds/cookoo/log"
+	"github.com/Masterminds/cookoo/safely"
+
+	"github.com/coreos/go-etcd/etcd"
+)
+
+var (
+	retryCycles = 2
+	retrySleep  = 200 * time.Millisecond
+)
+
+// Getter describes the Get behavior of an Etcd client.
+//
+// Usually you will want to use go-etcd/etcd.Client to satisfy this.
+//
+// We use an interface because it is more testable.
+type Getter interface {
+	Get(string, bool, bool) (*etcd.Response, error)
+}
+
+// DirCreator describes etcd's CreateDir behavior.
+//
+// Usually you will want to use go-etcd/etcd.Client to satisfy this.
+type DirCreator interface {
+	CreateDir(string, uint64) (*etcd.Response, error)
+}
+
+// Watcher watches an etcd entry.
+type Watcher interface {
+	Watch(string, uint64, bool, chan *etcd.Response, chan bool) (*etcd.Response, error)
+}
+
+// Setter sets a value in Etcd.
+type Setter interface {
+	Set(string, string, uint64) (*etcd.Response, error)
+}
+
+// GetterSetter performs get and set operations.
+type GetterSetter interface {
+	Getter
+	Setter
+}
+
+// CreateClient creates a new Etcd client and prepares it for work.
+//
+// Params:
+// 	- url (string): A server to connect to.
+// 	- retries (int): Number of times to retry a connection to the server
+// 	- retrySleep (time.Duration): How long to sleep between retries
+//
+// Returns:
+// 	This puts an *etcd.Client into the context.
+func CreateClient(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	url := p.Get("url", "http://localhost:4001").(string)
+
+	// Backed this out because it's unnecessary so far.
+	//hosts := p.Get("urls", []string{"http://localhost:4001"}).([]string)
+	hosts := []string{url}
+	retryCycles = p.Get("retries", retryCycles).(int)
+	retrySleep = p.Get("retrySleep", retrySleep).(time.Duration)
+
+	// Support `host:port` format, too.
+	for i, host := range hosts {
+		if !strings.Contains(host, "://") {
+			hosts[i] = "http://" + host
+		}
+	}
+
+	client := etcd.NewClient(hosts)
+	client.CheckRetry = checkRetry
+
+	return client, nil
+}
+
+// Get performs an etcd Get operation.
+//
+// Params:
+// 	- client (EtcdGetter): Etcd client
+// 	- path (string): The path/key to fetch
+//
+// Returns:
+// - This puts an `etcd.Response` into the context, and returns an error
+//   if the client could not connect.
+func Get(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	cli, ok := p.Has("client")
+	if !ok {
+		return nil, errors.New("No Etcd client found.")
+	}
+	client := cli.(Getter)
+	path := p.Get("path", "/").(string)
+
+	res, err := client.Get(path, false, false)
+	if err != nil {
+		return res, err
+	}
+
+	if !res.Node.Dir {
+		return res, fmt.Errorf("Expected / to be a dir.")
+	}
+	return res, err
```

Comment:
Shouldn't this be `return res, nil`? I know `err` is probably equal to `nil` if we got here, but go code tends to make it explicit IMHO.